File: codechef/feb18/lucasth.cc
// https://www.codechef.com/FEB18/problems/LUCASTH
#include <iostream>
#include <vector>

using namespace std;

typedef vector<int> vi;

void s(vi &a, int p) {
  int n = a.size();
  vi b(n, 0);
  a[0] = 1;
  for (int i = 1; i < n; i++) {
    for (int j = 1; j <= i; j++)
      b[j] = (a[j - 1] + (i - 1) * a[j]) % p;
    swap(a, b);
    a[0] = 0;
  }
}

int main() {
  int t;
  cin >> t;
  while (t--) {
    int n, p;
    cin >> n >> p;
    n += 2;
    vi a(n, 0);
    s(a, p);
    int c = 0;
    for (int i = 0; i < n; i++)
      if (a[i] % p) c++;
    cout << c << endl;
  }
}

File: codechef/feb18/lucasth.rs
// https://www.codechef.com/FEB18/problems/LUCASTH
use std::io;
use std::mem;

fn s(n: usize, p: usize) {
    let mut a = vec![0; n];
    let mut b = vec![0; n];
    a[0] = 1;
    for i in 1 .. n {
        for j in 1 .. i + 1 {
            b[j] = (a[j - 1] + (i - 1) * a[j]) % p;
        }
        b = mem::replace(&mut a, b);
        a[0] = 0;
    }
    let mut c = 0;
    for i in 0 .. n {
        if a[i] != 0 { c += 1 }
    }
    println!("{}", c);
}

fn main() {
    let stdin = io::stdin();
    let mut line = String::new();
    stdin.read_line(&mut line).unwrap();
    let t = line.trim().parse::<usize>().unwrap();
    for _ in 0..t {
        line.clear();
        stdin.read_line(&mut line).unwrap();
        let numbers = line
            .split_whitespace()
            .map(|x| x.parse::<usize>().unwrap())
            .collect::<Vec<usize>>();
        let n = numbers[0] + 2 as usize;
        let p = numbers[1];
        s(n, p);
    }
}
